File: frontend/src/types/history.ts
/* history.ts — types for /api/history/* responses (Stage 1 + 2). */

export interface RunSummary {
  id: number
  run_type: string
  status: 'running' | 'completed' | 'partial' | 'failed' | string
  started_at: string         // ISO 8601
  finished_at: string | null
  parsers: string[]
  files_count: number
}

export interface ResultRow {
  parser_name: string
  file_name: string
  coverage_checked: number
  coverage_failed: number
  coverage_rate: number
  noise_checked: number
  noise_failed: number
  noise_rate: number
  avg_score: number
  gt_word_count: number
  parser_word_count: number
  diagnostics_json: Record<string, unknown> | null
}

export interface RunDetail extends RunSummary {
  selected_files: string[]
  results: ResultRow[]
}

export interface TimelinePoint {
  run_id: number
  started_at: string
  parser_name: string
  coverage: number | null
  noise: number | null
  avg_score: number | null
  doc_count: number
  file_set_changed: boolean
}

export interface TimelineFilterEcho {
  parser: string
  files: string[] | null
}

export interface TimelineResponse {
  points: TimelinePoint[]
  filter_echo: TimelineFilterEcho
}

export interface OverallTimelinePoint {
  run_id: number
  started_at: string
  overall_score: number | null
  contributing_corpora: number
  ratios_sum: number
  normalization_applied: boolean
  warning: string | null
  per_corpus: Record<string, number | null>
}

export interface OverallTimelineResponse {
  points: OverallTimelinePoint[]
  parser: string
}

export type TimelineMode = 'all' | 'files' | 'corpus' | 'overall'
export type TimelineMetric = 'avg_score' | 'coverage' | 'noise'

File: backend/app/schemas/history.py
"""
backend/app/schemas/history.py
--------------------------------
Pydantic response contracts for /api/history/* endpoints.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ResultRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    parser_name: str
    file_name: str
    coverage_checked: int
    coverage_failed: int
    coverage_rate: float
    noise_checked: int
    noise_failed: int
    noise_rate: float
    avg_score: float
    gt_word_count: int
    parser_word_count: int
    diagnostics_json: dict[str, Any] | None = None


class RunSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_type: str
    status: str
    started_at: datetime
    finished_at: datetime | None
    parsers: list[str] = Field(default_factory=list)
    files_count: int = 0


class RunDetail(RunSummary):
    selected_files: list[str] = Field(default_factory=list)
    results: list[ResultRow] = Field(default_factory=list)


class TimelinePoint(BaseModel):
    run_id: int
    started_at: datetime
    parser_name: str
    coverage: float | None
    noise: float | None
    avg_score: float | None
    doc_count: int
    file_set_changed: bool


class TimelineFilterEcho(BaseModel):
    parser: str
    files: list[str] | None = None


class TimelineResponse(BaseModel):
    points: list[TimelinePoint] = Field(default_factory=list)
    filter_echo: TimelineFilterEcho
